File: native/ios/Sources/StingRuntime/StingPresentationHostView.swift
import UIKit

/// Invisible renderer node that owns content presented by UIKit.
///
/// The content stack survives dismiss/re-present cycles, preserving Solid/native
/// identity while the `presented` prop remains application-authoritative.
final class StingPresentationHostView: UIView, UIAdaptivePresentationControllerDelegate {
    let kind: String
    let contentStack = UIStackView()

    private var requestedPresented = false
    private weak var presentedController: UIViewController?
    private var dismissHandler: (() -> Void)?
    private var suppressDismissEvent = false

    init(frame: CGRect = .zero, kind: String) {
        precondition(kind == "modal" || kind == "sheet", "Unsupported presentation kind: \(kind)")
        self.kind = kind
        super.init(frame: frame)
        configure()
    }

    required init?(coder: NSCoder) {
        self.kind = "modal"
        super.init(coder: coder)
        configure()
    }

    private func configure() {
        contentStack.axis = .vertical
        contentStack.alignment = .fill
        contentStack.distribution = .fill
        contentStack.spacing = 0
        isHidden = true
    }

    func setPresented(_ presented: Bool) {
        requestedPresented = presented
        if presented {
            presentIfPossible()
        } else {
            dismiss(programmatic: true)
        }
    }

    func setDismissHandler(enabled: Bool, handler: @escaping () -> Void) {
        dismissHandler = enabled ? handler : nil
    }

    override func didMoveToWindow() {
        super.didMoveToWindow()
        if window != nil, requestedPresented {
            presentIfPossible()
        } else if window == nil {
            dismiss(programmatic: true)
        }
    }

    func disposePresentation() {
        requestedPresented = false
        dismiss(programmatic: true)
        dismissHandler = nil
    }

    var isRequestedPresentedForTesting: Bool { requestedPresented }

    func simulateDismissForTesting() {
        guard requestedPresented else { return }
        requestedPresented = false
        dismissHandler?()
    }

    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
        guard !suppressDismissEvent else { return }
        requestedPresented = false
        presentedController = nil
        dismissHandler?()
    }

    private func presentIfPossible() {
        guard requestedPresented,
              presentedController == nil,
              let presenter = nearestViewController(),
              presenter.presentedViewController == nil else { return }

        let controller = UIViewController()
        controller.view.backgroundColor = kind == "sheet" ? .systemBackground : .clear
        contentStack.removeFromSuperview()
        contentStack.translatesAutoresizingMaskIntoConstraints = false
        controller.view.addSubview(contentStack)
        NSLayoutConstraint.activate([
            contentStack.leadingAnchor.constraint(equalTo: controller.view.leadingAnchor),
            contentStack.trailingAnchor.constraint(equalTo: controller.view.trailingAnchor),
            contentStack.topAnchor.constraint(equalTo: controller.view.topAnchor),
            contentStack.bottomAnchor.constraint(equalTo: controller.view.bottomAnchor),
        ])

        controller.modalPresentationStyle = kind == "sheet" ? .pageSheet : .overFullScreen
        presentedController = controller
        presenter.present(controller, animated: false) { [weak controller, weak self] in
            controller?.presentationController?.delegate = self
        }
    }

    private func dismiss(programmatic: Bool) {
        guard let controller = presentedController else { return }
        suppressDismissEvent = programmatic
        controller.dismiss(animated: false) { [weak self] in
            guard let self else { return }
            self.contentStack.removeFromSuperview()
            self.presentedController = nil
            self.suppressDismissEvent = false
        }
    }

    private func nearestViewController() -> UIViewController? {
        var responder: UIResponder? = self
        while let current = responder {
            if let controller = current as? UIViewController { return controller }
            responder = current.next
        }
        return window?.rootViewController
    }
}

File: native/android/src/main/java/run/stingjs/runtime/StingPresentationHostView.kt
package run.stingjs.runtime

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.Gravity
import android.view.ViewGroup
import android.view.Window
import android.widget.FrameLayout
import android.widget.LinearLayout

/**
 * Invisible renderer node that owns content presented through an Android Dialog.
 * The content container survives dismiss/re-present cycles so Solid/native identity
 * remains stable while presentation is controlled declaratively.
 */
internal class StingPresentationHostView(
    context: Context,
    val kind: String,
) : FrameLayout(context) {
    val content = LinearLayout(context).apply {
        orientation = LinearLayout.VERTICAL
    }

    private var requestedPresented = false
    private var dialog: Dialog? = null
    private var dismissHandler: (() -> Unit)? = null
    private var suppressDismissEvent = false

    init {
        require(kind == "modal" || kind == "sheet") { "Unsupported presentation kind: $kind" }
        layoutParams = ViewGroup.LayoutParams(0, 0)
    }

    fun setPresented(presented: Boolean) {
        requestedPresented = presented
        if (presented) showIfPossible() else dismiss(programmatic = true)
    }

    fun setDismissHandler(enabled: Boolean, handler: () -> Unit) {
        dismissHandler = if (enabled) handler else null
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (requestedPresented) showIfPossible()
    }

    override fun onDetachedFromWindow() {
        dismiss(programmatic = true)
        super.onDetachedFromWindow()
    }

    fun disposePresentation() {
        requestedPresented = false
        dismiss(programmatic = true)
        dismissHandler = null
    }

    internal fun isPresentedForTesting(): Boolean = requestedPresented

    internal fun simulateDismissForTesting() {
        if (!requestedPresented) return
        requestedPresented = false
        dismissHandler?.invoke()
    }

    private fun showIfPossible() {
        if (!isAttachedToWindow || dialog?.isShowing == true) return
        val activity = context.findActivity() ?: return
        if (content.parent != null) {
            (content.parent as? ViewGroup)?.removeView(content)
        }

        val next = Dialog(activity).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(content)
            setOnDismissListener {
                val shouldEmit = !suppressDismissEvent && requestedPresented
                requestedPresented = false
                dialog = null
                if (shouldEmit) dismissHandler?.invoke()
            }
            window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            window?.setLayout(
                ViewGroup.LayoutParams.MATCH_PARENT,
                if (kind == "sheet") ViewGroup.LayoutParams.WRAP_CONTENT else ViewGroup.LayoutParams.MATCH_PARENT,
            )
            if (kind == "sheet") {
                window?.setGravity(Gravity.BOTTOM)
            }
        }
        dialog = next
        next.show()
        next.window?.setLayout(
            ViewGroup.LayoutParams.MATCH_PARENT,
            if (kind == "sheet") ViewGroup.LayoutParams.WRAP_CONTENT else ViewGroup.LayoutParams.MATCH_PARENT,
        )
        if (kind == "sheet") next.window?.setGravity(Gravity.BOTTOM)
    }

    private fun dismiss(programmatic: Boolean) {
        val current = dialog ?: return
        suppressDismissEvent = programmatic
        try {
            current.dismiss()
        } finally {
            suppressDismissEvent = false
            dialog = null
        }
    }
}

private fun Context.findActivity(): Activity? {
    var current: Context? = this
    while (current is ContextWrapper) {
        if (current is Activity) return current
        current = current.baseContext
    }
    return current as? Activity
}
